### PR TITLE
Added headings option to control optional headings, changed all headings to h1, #2729

### LINKED ISSUE #2729: Better page heading behavior
## Search Terms

readme index heading

## Problem

Issue 1:
All other pages, excluding modules.html (see #2728), have an h1 heading, while the readme index.html page has an h2 heading. This causes an inconsistent look.

Issue 2:
Many (probably most) typescript projects start their README.md files with a project name heading, which leads to redundant headings when combined with the readme index.html page title.

## Suggested Solution

Issue 1:
Change the index.html heading from h2 to h1.

Issue 2 Option 1:
Add an `omitReadmeHeading` boolean option that skips rendering the page title heading.

Issue 2 Option 2:
Add an `omitReadmeHeading` boolean option that checks if the README.md file starts with a heading that matches the project name and removes it if found. This would have the added benefit of simplifying the `On This Page` section.

Issue 2 Option 3:
Same as option 2 but have the `omitReadmeHeading` option default to `true`.

Issue 2 Option 4 (preferred):
Check if the README.md file starts with a heading that matches the project name and remove it if found by default. This would have the added benefits of simplifying the `On This Page` section and not adding more options.

### PR DESCRIPTION
**Pull request by mrfigg**
Resolves #2729.

### CODE REVIEW

File: src/lib/output/themes/default/partials/header.tsx [hunk 1]
```diff
@@ -5,12 +5,13 @@ import type { PageEvent } from "../../../events";
...
-    const HeadingLevel = props.model.isProject() ? "h2" : "h1";
+    const opts = context.options.getValue("headings");
+
     return (
         <div class="tsd-page-title">
             {!!props.model.parent && <ul class="tsd-breadcrumb">{context.breadcrumb(props.model)}</ul>}
-            {!props.model.isDocument() && (
-                <HeadingLevel class={classNames({ deprecated: props.model.isDeprecated() })}>
+            {(!props.model.isProject() || opts.project) && (!props.model.isDocument() || opts.document) && (
+                <h1 class={classNames({ deprecated: props.model.isDeprecated() })}>
```
Comment (Gerrit0) [inline src/lib/output/themes/default/partials/header.tsx:14]: We should make the change to h1 in hierarchy.tsx as well

File: src/lib/output/themes/default/partials/header.tsx [hunk 2]
```diff
@@ -5,12 +5,13 @@ import type { PageEvent } from "../../../events";
...
 export const header = (context: DefaultThemeRenderContext, props: PageEvent<Reflection>) => {
-    const HeadingLevel = props.model.isProject() ? "h2" : "h1";
+    const opts = context.options.getValue("headings");
+
     return (
         <div class="tsd-page-title">
             {!!props.model.parent && <ul class="tsd-breadcrumb">{context.breadcrumb(props.model)}</ul>}
-            {!props.model.isDocument() && (
-                <HeadingLevel class={classNames({ deprecated: props.model.isDeprecated() })}>
+            {(!props.model.isProject() || opts.project) && (!props.model.isDocument() || opts.document) && (
```
Comment (Gerrit0) [inline src/lib/output/themes/default/partials/header.tsx:13]: Hmmmm.... I've realized that this is actually kind of problematic due to the readme. Both the index and the root modules pages are rendered with the project as the model.... and it's probably somewhat confusing that changing the project option will affect both of the pages... 

Not quite sure how we ought to handle this yet... going to sleep on it.

Comment (mrfigg) [reply src/lib/output/themes/default/partials/header.tsx:13]: Yea, I agree, not sure how that should be handled. We could add a `headings.readme` flag and then in `header.tsx` use something like the following:

```tsx
//...
{
    (
        !props.model.isProject() ||
        (props.url === "index.html" && opts.readme) ||
        (props.url !== "index.html" && opts.project)
    ) &&
    (!props.model.isDocument() || opts.document) &&
    (<h1...> ... </h1>)
}
// ...
```

Comment (Gerrit0) [reply src/lib/output/themes/default/partials/header.tsx:13]: On second thought... the project page is really just another module, so this check should probably just be for index.html, I think we should always include the title on the modules page.

**Comment by Gerrit0** [comment]
Thanks!

**Pull request merged**